internal/action/workspace.go

```diff
@@ -143,15 +144,62 @@ func NewWorkspaceResource(ctx context.Context, client *tfe.Client, workspaces []
 	ws.SSHKeyID = config.SSHKeyID
 	ws.WorkingDirectory = config.WorkingDirectory
 
+	if err := SetTags(ws, config.Tags); err != nil {
+		return nil, err
+	}
+
 	return ws, nil
 }
 
+type Tag string
+type Tags []Tag
+
 type TeamDataResource struct {
 	ForEach      map[string]TeamDataResource `json:"for_each,omitempty"`
 	Name         string                      `json:"name"`
 	Organization string                      `json:"organization"`
 }
 
+// SetTags adds workspace tags to the passed module
+func SetTags(module *tfeprovider.Workspace, tags map[string]Tags) error {
+	if len(tags) == 0 {
+		return nil
+	}
+
+	b, err := json.Marshal(tags)
+	if err != nil {
+		return fmt.Errorf("failed to marshal workspace tags: %s", err)
+	}
+
+	module.TagNames = fmt.Sprintf("${toset(lookup(%s, each.value.name, []))}", string(b))
+
+	return nil
+}
+
+// FormatTagsByWorkspace returns a map of tags by workspace
+func FormatTagsByWorkspace(tags Tags, wsTags map[string]Tags, workspaces []*Workspace) (map[string]Tags, error) {
+	tagsByWorkspace := map[string]Tags{}
+
+	if len(tags) == 0 && len(wsTags) == 0 {
```

Comment:
Returning early on empty lists is generally not necessary if code is not extremely performance sensitive. It's also possible that compiler optimizations make this no different than just writing the loop.